.github/workflows/daily-tests.yaml

```diff
@@ -8,404 +8,148 @@ on:
+    - uses: actions/checkout@v2
+    - id: artifact-name
+      run: echo "name=$(date +"%Y-%m-%d_%H.%M.%S-")" >> $GITHUB_OUTPUT
 
-  build-gem5-ALL_MSI:
+  build-gem5:
+    strategy:
+      matrix:
+        image: [ALL, ALL_CHI, ARM, ALL_MSI, ALL_MESI_Two_Level, NULL_MI_example, RISCV, VEGA_X86]
+        include:
+          - command-line: -j $(nproc)
+          - image: ALL_CHI
+            command-line: --default=ALL PROTOCOL=CHI -j $(nproc)
+          - image: ALL_MSI
+            command-line: --default=ALL PROTOCOL=MSI -j $(nproc)
+          - image: ALL_MESI_Two_Level
+            command-line: --default=ALL PROTOCOL=MESI_Two_Level -j $(nproc)
+          - image: NULL_MI_example
+            command-line: --default=NULL PROTOCOL=MI_example -j $(nproc)
     runs-on: [self-hosted, linux, x64, build]
+    needs: name-artifacts
     container: gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest
-    outputs:
-      build-name: ${{ steps.artifact-name.outputs.name }}
-    steps:
-      - uses: actions/checkout@v3
-        with:
-          # Scheduled workflows run on the default branch by default. We
-          # therefore need to explicitly checkout the develop branch.
-          ref: develop
-      - id: artifact-name
-        run: echo "name=$(date +"%Y-%m-%d_%H.%M.%S")-ALL_MSI" >> $GITHUB_OUTPUT
-      - name: Build gem5
-        run: |
-          scons build/ALL_MSI/gem5.opt --default=ALL PROTOCOL=MSI -j $(nproc)
-      - uses: actions/upload-artifact@v3
-        with:
-          name: ${{ steps.artifact-name.outputs.name }}
-          path: build/ALL_MSI/gem5.opt
-          retention-days: 5
-      - run: echo "This job's status is ${{ job.status }}."
-
-  build-gem5-ALL_MESI_Two_Level:
-    runs-on: [self-hosted, linux, x64, build,]
-    container: gcr.io/gem5-test/ubuntu-22.04_all-dependencies:latest
-    outputs:
-      build-name: ${{ steps.artifact-name.outputs.name }}
     steps:
       - uses: actions/checkout@v3
         with:
           # Scheduled workflows run on the default branch by default. We
           # therefore need to explicitly checkout the develop branch.
           ref: develop
-      - id: artifact-name
-        run: echo "name=$(date +"%Y-%m-%d_%H.%M.%S")-ALL_MESI_Two_Level" >> $GITHUB_OUTPUT
       - name: Build gem5
-        run: |
-          scons build/ALL_MESI_Two_Level/gem5.opt --default=ALL PROTOCOL=MESI_Two_Level -j $(nproc)
+        run: scons build/${{ matrix.image }}/gem5.opt ${{ matrix.command-line }}
       - uses: actions/upload-artifact@v3
         with:
-          name: ${{ steps.artifact-name.outputs.name }}
-          path: build/ALL_MESI_Two_Level/gem5.opt
+          name: ${{ needs.name-artifacts.outputs.build-name }}${{ matrix.image }}
+          path: build/${{ matrix.image }}/gem5.opt
           retention-days: 5
       - run: echo "This job's status is ${{ job.status }}."
 
+  # NULL build added separately since NULL is a keyword in yaml files
   build-gem5-NULL:
```

Comment:
You explained to be earlier that you separated this out to it's own separate build stop because "NULL" is keyword. Are you sure you can't just put this as a string in the `image` matrix? E.g., `image: ["NULL", "ALL"...`.